Add an event so that users can detect when an Application icon is clicked

danwalmsley opened the pull request:
On OSX when using `ShutdownMode.OnExplicitShutdown`, and the application has all its windows closed, the application icon remains in the Dock.

Most end-users expect that clicking the Dock icon will re-open the main window of the application. (See safari and other similar apps on MacOS).

However since there is no way for an Avalonia application to detect or handle this event, Avalonia applications can not implement this behaviour.

Applications can also not detect when the Dock icon is right clicked-> Hide or right clicked and Show... where the application is placed into a background state on macOS.

~~This PR adds an `Application.Clicked` event which only gets triggered on MacOS when the Dock icon (which represents the `Application` and not a `Window`) is clicked.~~

This PR adds an `IActivatableLifetime` that provides `Activated` and `Deactivated` events with event args providing the reason.

This api covers all bases for the specified scenarios on OSX and the api can be used as follows:

```
public override void OnFrameworkInitializationCompleted()
        {
            if (ApplicationLifetime is IClassicDesktopStyleApplicationLifetime desktopLifetime)
            {
                desktopLifetime.MainWindow = new MainWindow();
                desktopLifetime.ShutdownMode = ShutdownMode.OnExplicitShutdown;
            }

            if (ApplicationLifetime is IActivatableApplicationLifetime activatableApplicationLifetime)
            {
                activatableApplicationLifetime.Activated += ActivatableApplicationLifetimeOnActivated;
                
                activatableApplicationLifetime.Deactivated += ActivatableApplicationLifetimeOnDeactivated;
            }
        }

        private void ActivatableApplicationLifetimeOnActivated(object sender, ActivatedEventArgs e)
        {
            if (e.Kind == ActivationKind.Reopen)
            {
                if (ApplicationLifetime is IClassicDesktopStyleApplicationLifetime desktopLifetime)
                {
                    if (desktopLifetime.Windows.Count == 0)
                    {
                        desktopLifetime.MainWindow = new MainWindow();
                        desktopLifetime.MainWindow.Show();
                    }
                }
            }
            else if (e.Kind == ActivationKind.OpenUri && e is ProtocolActivatedEventArgs pargs)
            {
                Console.WriteLine($"Open uri {pargs.Uri}");
            }
            else if (e.Kind == ActivationKind.Background)
            {
                Console.WriteLine("Unhidden");
            }
        }
        
        private void ActivatableApplicationLifetimeOnDeactivated(object sender, ActivatedEventArgs e)
        {
            if (e.Kind == ActivationKind.Background)
            {
                Console.WriteLine("Hidden");
            }
        }
```

This allows the developer to handle this in any way and implement the behaviour they wish for their application.

Commit:
Add an event so that users can detect when an Application icon is clicked.

thevortexcloud commented:
> This PR adds an Application.Clicked

Should it be called something else so it's consistent with the rest of the Avalonia events? EG `PointerPressed`

maxkatz6 commented:
I think we are good with one of two possible APIs:
1. Something that follows UWP-like API:
```
public enum ActivationKind
{
    Reopen = 10, // Dock?
}
public class ActivatedEventArgs : EventArgs
{
    public ActivationKind Kind {get;}
}

public class Application
{
    public event EventHandler<ActivatedEventArgs> Activated;
}
```

2. Or `MacOS.HandleReopenEvent` attached event, specific to MacOS. And following https://github.com/AvaloniaUI/Avalonia/issues/11384. This attached event also wouldn't be a big API blocker later.

workgroupengineering commented:
> I think we are good with one of two possible APIs:
> 
>     1. Something that follows UWP-like API:
> 
> 
> ```
> public enum ActivationKind
> {
>     Reopen = 10, // Dock?
> }
> public class ActivatedEventArgs : EventArgs
> {
>     public ActivationKind Kind {get;}
> }
> 
> public class Application
> {
>     public event EventHandler<ActivatedEventArgs> Activated;
> }
> ```
> 
>     2. Or `MacOS.HandleReopenEvent` attached event, specific to MacOS. And following [Platform-specific APIs. #11384](https://github.com/AvaloniaUI/Avalonia/issues/11384). This attached event also wouldn't be a big API blocker later, if we decide to use first approach instead at some point.

I think it's the best choice, I think we should also migrate [IApplicationPlatformEvents](https://github.com/AvaloniaUI/Avalonia/blob/38e839997d2c204548e6fad396c178780a010cb1/src/Avalonia.Controls/Platform/IApplicationPlatformEvents.cs).

maxkatz6 commented:
@workgroupengineering yes, it would make sense there too

danwalmsley commented:
@maxkatz6 @workgroupengineering, had an offline discussion with @maxkatz6, 

will implement the above suggestion, but making it part of the application lifetime interface.

Commit:
refactor to use Lifetime apis.

Commit:
Merge remote-tracking branch 'origin/master' into features/osx-allow-app-to-reopen-main-window

Commit:
use ActivationKind instead of reason to be consistent with other xaml platforms

Commit:
implement macos raise url events.

danwalmsley commented:
Refactored using lifetimes and updated the description to show api usage.

maxkatz6 commented on src/Avalonia.Controls/ApplicationLifetimes/IActivatableApplicationLifetime.cs:
```diff
@@ -0,0 +1,10 @@
+{
+    event EventHandler<ActivatedEventArgs> Activated;
+    
+    event EventHandler<ActivatedEventArgs> Deactivated;
```
I feel like we should split these args, and add DeactivatedEventArgs instead of reusing. Maybe even with DeactivationKind.

workgroupengineering replied:
If I interpreted correctly what @danwalmsley meant if `ActivationKind.Background` occurs in the Activated event, it indicates that the app has been reactivated from suspension, otherwise in Deactivated it indicates that the app has been suspended. So for each of the elements of the enum.

If so I would suggest renaming:
  ActivatedEventArgs in ApplicationStateEventArgs
  ActivationKind in ApplicationStateKind

and create

  ApplicationStateActivatedEventArgs: ApplicationStateEventArgs
  ApplicationStateDeactivatedEventArgs: ApplicationStateEventArgs

cc @robloo

maxkatz6 commented on src/Avalonia.Controls/ApplicationLifetimes/ActivationKind.cs:
```diff
@@ -0,0 +1,8 @@
+
+public enum ActivationKind
+{
+    OpenUri = 20,
+    Reopen = 30, // i.e. Dock Icon clicked on MacOS
+    Background = 40 // i.e. Entered or left background.
```
nit: doc comments here and on other new public types

danwalmsley replied:
commented public api.

Commit:
add docs.

Commit:
add apis to programatically Activate and Deactivate the application.

This allows the dock icon to be kept in sync so its menu options there say "Hide" / "Show" correctly.

Commit:
Merge branch 'master' into features/osx-allow-app-to-reopen-main-window

Commit:
fix api naming.

maxkatz6 committed:
Add Browser IActivatableApplicationLifetime impl

maxkatz6 committed:
Implement IActivatableApplicationLifetime on Android

maxkatz6 committed:
Add IActivatableApplicationLifetime iOS implementation

maxkatz6 committed:
Adjust android impl a little

maxkatz6 committed:
Merge branch 'master' into features/osx-allow-app-to-reopen-main-window
